### ISSUE TITLE
GoAgent connection issue with java.lang.NoClassDefFoundError:javax.crypto.JceSecurity error

### ISSUE THREAD

**Issue by rahulghanate** [gocd/gocd#3023]
##### Issue Type
- Bug Report

##### Summary

GoAgent on a specific host is not able to communicate with GoServer with below exception in logs,
```
2016-12-21 10:34:37,674 [loopThread] ERROR springframework.scheduling.timer.MethodInvokingTimerTaskFactoryBean:68 - Invocation of method 'loop' on target class [class com.thoughtworks.go.agent.AgentController] failed
java.lang.NoClassDefFoundError: Could not initialize class javax.crypto.JceSecurity
        at javax.crypto.KeyAgreement.getInstance(KeyAgreement.java:173)
        at sun.security.ssl.JsseJce.getKeyAgreement(JsseJce.java:285)
        at sun.security.ssl.JsseJce.isEcAvailable(JsseJce.java:198)
        at sun.security.ssl.CipherSuite$KeyExchange.isAvailable(CipherSuite.java:376)
        at sun.security.ssl.CipherSuite.isAvailable(CipherSuite.java:194)
        at sun.security.ssl.SSLContextImpl.getApplicableCipherSuiteList(SSLContextImpl.java:338)
        at sun.security.ssl.SSLContextImpl.getDefaultCipherSuiteList(SSLContextImpl.java:296)
        at sun.security.ssl.SSLSocketImpl.init(SSLSocketImpl.java:583)
```

I also sometimes get logs as below,
```
2016-12-21 10:25:04,612 [loopThread] ERROR springframework.scheduling.timer.MethodInvokingTimerTaskFactoryBean:68 - Invocation of method 'loop' on target class [class com.thoughtworks.go.agent.AgentController] failed
java.lang.OutOfMemoryError: GC overhead limit exceeded
        at java.io.UnixFileSystem.list(Native Method)
        at java.io.File.list(File.java:1116)
        at javax.crypto.JarVerifier.getSystemEntropy(JarVerifier.java:788)
        at javax.crypto.JarVerifier.testSignatures(JarVerifier.java:706)
        at javax.crypto.JarVerifier.access$400(JarVerifier.java:34)
        at javax.crypto.JarVerifier$1.run(JarVerifier.java:183)
        at javax.crypto.JarVerifier$1.run(JarVerifier.java:149)
```

##### Environment

Go-Server
Ubuntu 12.04 with 32GB RAM
Go-Agent
Ubuntu 12.04 with 32 GB RAM
Current memory snapshot of free command
```
# free -m
             total       used       free     shared    buffers     cached
Mem:         32040      16638      15402          0        836       9833
-/+ buffers/cache:       5968      26072
Swap:        16375       3382      12993
```
GoAgent process details,
```
go       18174  1.1  1.9 10197348 648984 pts/31 Sl  10:24   0:09 /usr/lib/jvm/jdk1.7.0_79/jre//bin/java -jar /usr/share/go-agent/agent-bootstrapper.jar -serverUrl https://go-server.example.com:8154/go
go       18198  6.0  1.1 2099204 372896 pts/31 Sl   10:24   0:50 /usr/lib/jvm/jdk1.7.0_79/jre/bin/java -Dcruise.console.publish.interval=10 -Xms128m -Xmx256m -Djava.security.egd=file:/dev/./urandom -Dagent.launcher.version=16.7.0-3819 -Dagent.plugins.md5=5a976380b8afb341e065bb3be9984a00 -Dagent.binary.md5=fnAwazgBWpNX7Wgcjh2Cow== -Dagent.launcher.md5=7HM5/ohlBdv+yi2gfQhhqQ== -jar agent.jar -serverUrl https://go-server.example.com:8154/go -sslVerificationMode NONE
```
For stackoverflow I came to know  local_policy.jar and US_export_policy.jar, should not be corrupted. But the agent machine also has the valid security libs. To clear doubt I even downloaded latest files from 
http://www.oracle.com/technetwork/java/javase/downloads/jce-7-download-432124.html
```
/usr/lib/jvm/jdk1.7.0_79/jre/lib/security # ls -ltr
total 144
-rw-r--r-- 1 uucp 143   158 Mar 16  2015 javafx.policy
-rw-r--r-- 1 uucp 143  2487 Apr 10  2015 US_export_policy.jar
-rw-r--r-- 1 uucp 143  2971 Apr 10  2015 local_policy.jar
-rw-r--r-- 1 uucp 143     0 Apr 10  2015 trusted.libraries
-rw-r--r-- 1 uucp 143 18033 Apr 10  2015 java.security
-rw-r--r-- 1 uucp 143  2593 Apr 10  2015 java.policy
-rw-r--r-- 1 uucp 143 98626 Apr 10  2015 cacerts
-rw-r--r-- 1 uucp 143  4054 Apr 10  2015 blacklist
-rw-r--r-- 1 uucp 143    98 Apr 10  2015 javaws.policy
```

###### Basic environment details

* Go Version: Go Version: 16.7.0
* JAVA Version: JVM version:	1.7.0_91
* OS: Linux 3.2.0-80-generic
* Browser vendor and version (if relevant): `Hopefully not Internet Explorer 7 :)`

##### Steps to Reproduce

1. Extracted go-agent 16.7 installation zip to /var/lib/
2. Created link to /var/lib/go-agent to /usr/share/
3. start go-agent with service go-agent command, which eventually runs agent.sh
4. Analyzed all logs in /var/log but the errors are not shown in any of those, but instead in /var/lib/go-agent/go-agent.log

##### Expected Results
Go-Agent starts smoothly and connects go-server

##### Actual Results
go-server Agents page doesn't show the specific host registered in the list.

**Comment by ketan** [comment]
Can you provide the JVM details on the server along with the api support output `https://go-server/go/api/support`

**Comment by rahulghanate** [comment]
The JVM installed on go-server is openjdk-7-jdk (installed with apt-get, /usr/lib/jvm/java-7-openjdk-amd64/jre/)
And the JVM installed on go-agent where it is causing the problem is jdk1.7.0_79 (To avoid disruption with default java on agent machine it's installed by extraction of zip file and added JAVA_HOME in /etc/default/go-agent file)
```
export GO_JAVA_HOME="/usr/lib/jvm/jdk1.7.0_79/jre/"
export JAVA_HOME="/usr/lib/jvm/jdk1.7.0_79/jre/" # SET_BY_GO_INSTALLER__DONT_REMOVE
```

Attached the file with output of api-support.
[go-server-api-support.txt](https://github.com/gocd/gocd/files/666344/go-server-api-support.txt)


Also here is the complete stack trace seen in /var/lib/go-agent/go-agent.log file,
```
2016-12-21 12:49:34,147 [loopThread] ERROR springframework.scheduling.timer.MethodInvokingTimerTaskFactoryBean:68 - Invocation of method 'loop' on target class [class com.thoughtworks.go.agent.AgentController] failed
java.lang.NoClassDefFoundError: Could not initialize class javax.crypto.JceSecurity
        at javax.crypto.KeyAgreement.getInstance(KeyAgreement.java:173)
        at sun.security.ssl.JsseJce.getKeyAgreement(JsseJce.java:285)
        at sun.security.ssl.JsseJce.isEcAvailable(JsseJce.java:198)
        at sun.security.ssl.CipherSuite$KeyExchange.isAvailable(CipherSuite.java:376)
        at sun.security.ssl.CipherSuite.isAvailable(CipherSuite.java:194)
        at sun.security.ssl.SSLContextImpl.getApplicableCipherSuiteList(SSLContextImpl.java:338)
        at sun.security.ssl.SSLContextImpl.getDefaultCipherSuiteList(SSLContextImpl.java:296)
        at sun.security.ssl.SSLSocketImpl.init(SSLSocketImpl.java:583)
        at sun.security.ssl.SSLSocketImpl.<init>(SSLSocketImpl.java:547)
        at sun.security.ssl.SSLSocketFactoryImpl.createSocket(SSLSocketFactoryImpl.java:109)
        at org.apache.http.conn.ssl.SSLConnectionSocketFactory.createLayeredSocket(SSLConnectionSocketFactory.java:363)
        at org.apache.http.conn.ssl.SSLConnectionSocketFactory.connectSocket(SSLConnectionSocketFactory.java:353)
        at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:141)
        at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:353)
        at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:380)
        at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:236)
        at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:184)
        at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88)
        at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
        at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:107)
        at com.thoughtworks.go.agent.common.ssl.GoAgentServerHttpClient.execute(GoAgentServerHttpClient.java:47)
        at com.thoughtworks.go.agent.service.AgentUpgradeService.checkForUpgrade(AgentUpgradeService.java:53)
        at com.thoughtworks.go.agent.service.AgentUpgradeService.checkForUpgrade(AgentUpgradeService.java:47)
        at com.thoughtworks.go.agent.AgentController.rpcLoop(AgentController.java:180)
        at com.thoughtworks.go.agent.AgentController.loop(AgentController.java:173)
        at sun.reflect.GeneratedMethodAccessor12.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.springframework.util.MethodInvoker.invoke(MethodInvoker.java:273)
        at org.springframework.scheduling.support.MethodInvokingRunnable.run(MethodInvokingRunnable.java:65)
        at org.springframework.scheduling.timer.DelegatingTimerTask.run(DelegatingTimerTask.java:70)
        at java.util.TimerThread.mainLoop(Timer.java:555)
        at java.util.TimerThread.run(Timer.java:505)
```

**Comment by arvindsv** [comment]
My guess is that you're missing the JCE policy files, which are usually installed as a part of the RPM/DEB installers.

http://www.oracle.com/technetwork/java/javase/downloads/jce8-download-2133166.html

http://stackoverflow.com/questions/5885639/java-security-class-couldnt-be-found

**Comment by rahulghanate** [comment]
Thanks Arvind for reply. I had already checked and  given those policy file details while filling the bug. 
The bug details also has the download link you suggested.

**Comment by arvindsv** [comment]
Oops. Sorry. I should have seen that. I just saw your latest reply. :(

**Comment by arvindsv** [comment]
Fine. :) My next guess is a mismatch between the Java you're using (in PATH) and the JAVA_HOME. You mentioned that JAVA_HOME is set to /usr/lib/jvm/jdk1.7.0_79/jre/, but your /go/api/support output is filled with /usr/lib/jvm/java-7-openjdk-amd64 as the path it's looking for libs from.

So, the Java you're running is not /usr/lib/jvm/jdk1.7.0_79/jre/bin/java, which is what I think you should prepend to path in `/etc/default/go-server`.

**Comment by arvindsv** [comment]
Grr. Still no, I think. :( Agent is using the right Java.

**Comment by arvindsv** [comment]
Could be a mismatch between the agent and server JRE/JDKs. Agent seems to be 7u79 and server seems to be 7u91. I remember that they removed some crypto stuff as a part of 7u85.

**Comment by rahulghanate** [comment]
Thanks Arvind for investigating on this. 
I have more than 30 hosts configured with puppet based installations using same go-agent and Java 7u79. And surprisingly they are are working fine with this go-server. 
I even tried deleting all related agent/java directories and reinstalling them with puppet module, but these errors are not going.

**Comment by jyotisingh** [comment]
@rahulghanate - I suspect java installation to be somehow corrupted on the agent machine. 

You mentioned: 
```
the JVM installed on go-agent where it is causing the problem is jdk1.7.0_79 (To avoid disruption with default java on agent machine it's installed by extraction of zip file and added JAVA_HOME in /etc/default/go-agent file)
```
Is this how rest of your agents(which are working as expected) are setup too or is there anything different about this one? In order to debug this, could you remove off this installation of java and set it up manually once (not with puppet) to see if that makes any difference?

**Comment by rahulghanate** [comment]
Thanks Jyoti for the reply.
Yup, the same is been done for other agents, and I tried it installing manually as well.
But this gave me an idea to manually install it with deb file and do update-alternatives to actual java needed by other apps on that machine.
Will try that.

**Comment by rahulghanate** [comment]
Looks like there is no deb file available in Java downloads section for this installation.
But I tried with different Java7 version 80 and that also didn't helped.
I think that host might have been tried connecting go-server before with some different cert and now after reinstalling go-agent, go-server doesn't allowing it.

**Comment by rahulghanate** [comment]
This issue is resolved.
The main problem behind this issue was high number of files in /tmp directory.
There were millions of uncleaned files present in /tmp directory and because of this "GC overhead limit exceeded" exception was been reported.

I cleaned all old files in /tmp directory and the go-agent could able to connect to go-server.
There must be some bug in code where it gets the list of files /tmp directory while creating a new temp/lock file.
Identified the problem after investigating on "GC overhead limit exceeded" exception on StackOverflow.

Closing this issue now, but the code need to fixed where it lists all files in /tmp/
